Improve Epub Format and remove images from gallery that are outside t…

belldandu opened the pull request:
…he gallery.

Adds the enhancements #4 and #8

belldandu commented on plugin/js/EpubItemSupplier.js:
```diff
@@ -18,12 +18,13 @@ EpubItemSupplier.prototype.setChapters = function (chapters) {
 EpubItemSupplier.prototype.createXhtmlFileName = function (fileIndex) {
-    return "index_split_" + util.zeroPad(fileIndex) + ".html";
+    // Belldandu: For now only new format gets implemented because not sure where this files functions are even called and i dont want to accidentally break something.
+    return "OEBPS/Text/[" + util.zeroPad(fileIndex) + "]" + ".xhtml";
```
Chapter Name was not done here because i'm not even sure this function gets called so there is no title to send to it.
Instead it follows part of the new format which is ```[0000].xhtml``` and so on.

belldandu commented on plugin/js/parsers/BakaTsukiImageCollector.js:
```diff
@@ -111,7 +112,11 @@ ImageElementConverter.prototype.replaceWithImagePageUrl = function (images) {
     if (imageInfo != null && that.element.parentElement != null) {
         let newImage = imageInfo.createImageElement();
-        that.element.parentElement.replaceChild(newImage, that.element);
+        if(that.element.className === "thumb" && (imageInfo.isOutsideGallery || imageInfo.isCover)){
```
imageInfo.isCover is never set to true or grabbed anywhere. This behavior should be changed in the future so that isCover is set to true for whatever image the checkmark is on for "Set Cover".

belldandu commented:
@dteviot whenever your ready.

belldandu commented:
@dteviot got any comments you want to add?

belldandu commented:
The overall changes are the following:
- Epub Format now uses a more common structure making editing easier.
- Chapter splits when possible use number + 20 character restricted chapter name otherwise they just use number
- Images that appear elsewhere within the epub that is to be made are removed from the novel illustration.
- Listing for images as well as their caption text in Novel Illustrations has been removed as this prevented images from getting their own pages.

All changes have been tested and seen working before this push was made.

dteviot commented:
Sorry, today has been chaos, (and tomorrow will probably be just as bad) so I haven't had time to examine your submission in detail.
On first look, it's pretty good, although there's some minor points I'd raise.

- There's not really any need to put your name in comments.  Git will show who's to blame.
- BakaTsukiParser.prototype.stripGalleryBoxWidthStyle() should realy be renamed as it's no longer doing that.
- After moving the images out of the list, you could discard the entire list.  which means you no longer need to dump the gallerytext elements.
- In EpubItem.prototype.getZipHref() I'd move the logic for creating/truncating a filename into separate function.  Share with EpubItemSupplier.prototype.createXhtmlFileName(), also BakaTsukiImageInfo.prototype.makeZipHref()  maybe put in Util?
- that.element.className === "thumb" && (imageInfo.isOutsideGallery || imageInfo.isCover), in ImageElementConverter.prototype.replaceWithImagePageUrl() I'd pull into own function.
- Unit tests are broken.
- Stripping images is an optional feature.  That said, I have some ideas on how to handle that and will take care of it this weekend.

Should have time to examine fully day after tomorrow.

belldandu commented:
Unit tests are broken for me regardless saying everything is undefined. They where broken the first day i cloned this repo.

as for 

> After moving the images out of the list, you could discard the entire list. which means you no longer need to dump the gallerytext elements.

I tried that. It led to images just not existing at all because i cant go higher or lower than 1 level for whatever reason. listItem.parentNode.parentNode === null and listItem.firstChild.firstChild === null

Other than that i cleaned up the other stuff based on your recommendations and attempted to fix the Utests for BakaTsuki.

@dteviot

belldandu commented on plugin/js/Util.js:
```diff
@@ -45,9 +45,10 @@ var util = (function () {
         let newImage = doc.createElementNS("http://www.w3.org/2000/svg","image");
         svg.appendChild(newImage);
-        newImage.setAttribute("xlink:href", href);
+        newImage.setAttribute("xlink:href", ".." + href.substr(5));
```
replace "OEBPS" with "..". Only gets done when retrieving href for the html. Do not do this elsewhere or it will break a lot of things.

belldandu commented on plugin/js/EpubPacker.js:
```diff
@@ -131,7 +131,7 @@ EpubPacker.prototype = {
         let that = this;
         var item = that.createAndAppendChild(manifest, "item");
-        item.setAttribute("href", href);
+        item.setAttribute("href", href.substr(6));
```
In certain places (specifically the manifest only) having "OEBPS/" before each item causes the list index to be out of range because of where content.opf and toc.ncx is.

We remove "OEBPS/" Only when inserting chapter href's into the manifest and nowhere else. Doing this anywhere else breaks a lot of things.

belldandu commented:
Well i'm off to bed. I don't mind fixing the unittests however i need them to actually work first and not say this.
![UnitTests?](http://i.imgur.com/ehqrssp.png)

belldandu commented:
There we go fixed a bunch of unit test stuff for baka tsuki.
Gonna start working on the other ones now. @dteviot

belldandu commented:
Actually all the other errors are just "Access to restricted URI denied" @dteviot and i have no idea whats doing it so Have fun with that.

belldandu commented:
Seems like you have missing functions in the sonako branch and its trying to call them but can't. @dteviot

belldandu commented:
You seem to be mostly using ECMA6 so why not use Class and Class Extends instead of function and Object.create @dteviot ?

belldandu commented:
Because your current usage falls under "Bad practice: Extension of native prototypes"

Refer to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain#Bad_practice_Extension_of_native_prototypes

belldandu commented on unitTest/UtestEpubPacker.js:
```diff
@@ -200,7 +200,7 @@ test("makeCoverImageXhtmlFile", function (assert) {
                "<div class=\"svg_outer svg_inner\">" +
                     "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" height=\"100%\" width=\"100%\" version=\"1.1\" preserveAspectRatio=\"xMidYMid meet\" viewBox=\"0 0 400 200\">" +
-                        "<image xlink:href=\"cover.png\" height=\"200\" width=\"400\"/>" +
+                        "<image xlink:href=\"...png\" height=\"200\" width=\"400\" data-origin=\"http://dummy/cover.png\"/>" +
```
This only happens because your function throws an error saying its not implemented yet.

belldandu commented:
Fixed some more UnitTest Stuff that i noticed

belldandu commented:
The rest is up to you on whether you want to implement class and class extends

belldandu commented on plugin/js/parsers/BakaTsukiImageCollector.js:
```diff
@@ -111,7 +112,18 @@ ImageElementConverter.prototype.replaceWithImagePageUrl = function (images) {
     if (imageInfo != null && that.element.parentElement != null) {
         let newImage = imageInfo.createImageElement();
-        that.element.parentElement.replaceChild(newImage, that.element);
+        if(0 === 1){ // placeholder for some check if we should remove the duplicate images (currently resolves to false)
```
Some variable that tells the epub gen to remove duplicate images. 
Resolves to false atm for obvious reasons.

belldandu commented:
@dteviot There i fixed as much of the unit tests as i could and added the ability to choose whether or not to remove duplicate images. I also made it so that the cover image is also properly removed from the Novel Illustration if said option is enabled.

belldandu commented:
![removeduplicateimagesoption](https://cloud.githubusercontent.com/assets/5639384/16305194/c7ef01a0-3926-11e6-884e-2c2cb77c8381.png)

belldandu commented:
Upon closer inspection of the "Restricted Access uri" errors i noticed that this is in its own essense cross origin policy issues. You cannot access the sites you are trying to access because they are of a different origin then your current url.

I suggest changing this behavior so you don't have these issues.

dteviot commented:
@belldandu 
> Apparently there is no way to even meet the spec of CORS for files loaded from the file:// protocol as the origin will always be null

see: http://www.chrome-allow-file-access-from-file.com/

> You seem to be mostly using ECMA6 so why not use Class and Class Extends instead of function and Object.create @dteviot ?

Because I'm still learning Javascript, and am not yet familiar with Classes.

> Because your current usage falls under "Bad practice: Extension of native prototypes"
Um.. no.  the relevant text is "One mis-feature that is often used is to extend **Object.prototype or one of the other built-in** prototypes."
I'm using this pattern: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Introduction_to_Object-Oriented_JavaScript

belldandu, I'm going to ask you to please stop submitting stuff for a couple of days until I have time to go through, assess and merge what you've done.  I know you're eager to get the gallery formatting stuff in.  (I am too.)  But I'd like to avoid you having to rework anything else you're doing based on what I'm willing to accept.
I'd also like to co-ordinate our development tasks so we don't wind up trying to do the same things.

belldandu commented:
@dteviot Ok

belldandu commented:
```
    var getChapterTitle = function (title) {
        if(title) {
            title = title.replace(/([^a-z0-9_\- ]+)/gi, '');
            return (title.length > 20 ? title.substr(0, 20) + "..." : title);
        }
        return "";
    }

    var getImageName = function (page) {
        if(page){
            let name = page.split(/file:/gi)[1];
            if(name){
                let image = name.split(/\./gi)[0].replace(/([^a-z0-9_\- ]+)/gi, '');
                return (image.length > 20 ? image.substr(0, 20) + "..." : image);
            }
        }
        return "";
    }
```

toshiya44 commented:
This might be slightly off-topic, but can you write a piece javascript that saves images from a baka-tsuki page, and that can be run on Firefox [scratchpad](http://i.imgur.com/IeQ8xHm.png)? A bit of explanation on how it works will be good too.

dteviot commented:
@toshiya44

> can you write a piece javascript that saves images from a baka-tsuki page, and that can be run on Firefox scratchpad? 

I don't have time to do this at moment.
I suggest you have a look at these:
Using javascript to make HTTP requests http://www.html5rocks.com/en/tutorials/es6/promises/ 
Parsing HTML: http://www.codeproject.com/Articles/1060680/Web-to-EPUB-Extension-for-Chrome
